src/core/hle/service/lm/lm.cpp

```diff
@@ -30,54 +79,257 @@ class ILogger final : public ServiceFramework<ILogger> {
+        offset += sizeof(LogPacketHeader);
 
-            SCOPE_EXIT({ addr += length; });
+        LogPacketHeaderEntrty entry{
+            .pid = header.pid,
+            .tid = header.tid,
+            .severity = header.severity,
+            .verbosity = header.verbosity,
+        };
 
-            if (field == Field::Skip) {
-                continue;
+        if (True(header.flags & LogPacketFlags::Head)) {
+            std::vector<u8> tmp(data.size() - sizeof(LogPacketHeader));
+            std::memcpy(tmp.data(), data.data() + offset, tmp.size());
+            entries[entry] = std::move(tmp);
+        } else {
+            // Append to existing entry
+            if (!entries.contains(entry)) {
+                LOG_ERROR(Service_LM, "Log entry does not exist!");
+                return;
             }
+            std::vector<u8> tmp(data.size() - sizeof(LogPacketHeader));
 
-            std::vector<u8> data(length);
-            memory.ReadBlock(addr, data.data(), length);
-            fields.emplace(field, std::move(data));
+            auto& existing_entry = entries[entry];
+            const auto base = existing_entry.size();
+            existing_entry.resize(base + (data.size() - sizeof(LogPacketHeader)));
+            std::memcpy(existing_entry.data() + base, data.data() + offset,
+                        (data.size() - sizeof(LogPacketHeader)));
         }
 
-        manager.Log({header, std::move(fields)});
+        if (True(header.flags & LogPacketFlags::Tail)) {
+            auto it = entries.find(entry);
+            if (it == entries.end()) {
+                LOG_ERROR(Service_LM, "Log entry does not exist!");
+                return;
+            }
+            ParseLog(it->first, it->second);
+            entries.erase(it);
+        }
     }
 
     void SetDestination(Kernel::HLERequestContext& ctx) {
         IPC::RequestParser rp{ctx};
-        const auto destination = rp.PopEnum<DestinationFlag>();
-
-        LOG_DEBUG(Service_LM, "called, destination={:08X}", destination);
+        const auto log_destination = rp.PopEnum<LogDestination>();
 
-        manager.SetDestination(destination);
+        LOG_DEBUG(Service_LM, "called, destination={}", DestinationToString(log_destination));
+        destination = log_destination;
 
         IPC::ResponseBuilder rb{ctx, 2};
         rb.Push(RESULT_SUCCESS);
     }
 
-    Manager& manager;
-    Core::Memory::Memory& memory;
+    u32 ReadLeb128(const std::vector<u8>& data, std::size_t& offset) {
+        u32 result{};
+        u32 shift{};
+        do {
+            result |= (data[offset] & 0x7f) << shift;
+            shift += 7;
+            offset++;
+            if (offset >= data.size()) {
+                break;
+            }
+        } while ((data[offset] & 0x80) != 0);
+        return result;
+    }
+
+    std::string ReadString(const std::vector<u8>& data, std::size_t& offset, std::size_t length) {
+        std::string output{};
```

Comment:
```suggestion
        std::string output(length, '\0');
```

This removes the need for a separate resize call